modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/QueryRegistryImpl.java

```diff
@@ -21,50 +21,90 @@
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
-
-import org.apache.ignite.IgniteLogger;
+import java.util.stream.Collectors;
+import org.apache.calcite.util.Pair;
+import org.apache.ignite.cache.query.QueryCancelledException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryType;
+import org.apache.ignite.internal.processors.query.GridQueryCancel;
 import org.apache.ignite.internal.processors.query.RunningQuery;
+import org.apache.ignite.internal.processors.query.RunningQueryManager;
+import org.apache.ignite.internal.processors.query.calcite.util.AbstractService;
 import org.apache.ignite.internal.util.IgniteUtils;
 
 /**
  * Registry of the running queries.
  */
-public class QueryRegistryImpl implements QueryRegistry {
+public class QueryRegistryImpl extends AbstractService implements QueryRegistry {
     /** */
-    private final ConcurrentMap<UUID, RunningQuery> runningQrys = new ConcurrentHashMap<>();
+    private final ConcurrentMap<UUID, Pair<Long, RunningQuery>> runningQrys = new ConcurrentHashMap<>();
 
     /** */
-    private final IgniteLogger log;
+    protected final GridKernalContext kctx;
 
     /** */
-    public QueryRegistryImpl(IgniteLogger log) {
-        this.log = log;
+    public QueryRegistryImpl(GridKernalContext ctx) {
+        super(ctx);
+
+        kctx = ctx;
     }
 
     /** {@inheritDoc} */
     @Override public RunningQuery register(RunningQuery qry) {
-        return runningQrys.computeIfAbsent(qry.id(), k -> qry);
+        return runningQrys.computeIfAbsent(qry.id(), k -> {
+            if (!(qry instanceof RootQuery))
+                return Pair.of(-1L, qry);
```

Comment:
Perhaps `Pair.of(null, qry)` is better, since `unregister` don't try to do anything if qryId is null.

Reply:
null is always worse that value